File: src/test/java/com/lehvolk/xodus/repo/SmartSearchQueryParserTest.java
package com.lehvolk.xodus.repo;

import java.util.Collections;
import java.util.List;

import org.junit.Test;

import com.lehvolk.xodus.repo.SearchTerm.Range;
import com.lehvolk.xodus.repo.SearchTerm.SearchTermType;
import static org.junit.Assert.assertEquals;

public class SmartSearchQueryParserTest {

    private static final String FIRST_NAME = "firstName";
    private static final String LAST_NAME = "lastName";

    @Test
    public void testSimple() {
        List<SearchTerm<?>> result = parse("firstName='John' and lastName='McCain'");
        assertEquals(2, result.size());
        checkTerm(result.get(0), FIRST_NAME, "John", SearchTermType.VALUE);
        checkTerm(result.get(1), LAST_NAME, "McCain", SearchTermType.VALUE);
    }

    @Test
    public void testSingleParam() {
        List<SearchTerm<?>> result = parse("firstName='John'");
        assertEquals(1, result.size());
        checkTerm(result.get(0), FIRST_NAME, "John", SearchTermType.VALUE);
    }

    @Test
    public void testWithoutQuotes() {
        List<SearchTerm<?>> result = parse("firstName=John and lastName=McCain and age=43");
        assertEquals(3, result.size());
        checkTerm(result.get(0), FIRST_NAME, "John", SearchTermType.VALUE);
        checkTerm(result.get(1), LAST_NAME, "McCain", SearchTermType.VALUE);
        checkTerm(result.get(2), "age", "43", SearchTermType.VALUE);
    }

    @Test
    public void testLike() {
        List<SearchTerm<?>> result = parse("firstName=John and lastName~McCain");
        assertEquals(2, result.size());
        checkTerm(result.get(0), FIRST_NAME, "John", SearchTermType.VALUE);
        checkTerm(result.get(1), LAST_NAME, "McCain", SearchTermType.LIKE);
    }

    @Test
    public void testSingleParamsWithoutQuotes() {
        List<SearchTerm<?>> result = parse("firstName=John");
        assertEquals(1, result.size());
        checkTerm(result.get(0), FIRST_NAME, "John", SearchTermType.VALUE);
    }

    @Test
    public void testEscapingQuotes() {
        List<SearchTerm<?>> result = parse("firstName=John and lastName='Mc''Cain'");
        assertEquals(2, result.size());
        checkTerm(result.get(0), FIRST_NAME, "John", SearchTermType.VALUE);
        checkTerm(result.get(1), LAST_NAME, "Mc'Cain", SearchTermType.VALUE);
    }

    @Test
    public void testOmitAndIntoQuotes() {
        List<SearchTerm<?>> result = parse("firstName='John and Mike' and lastName='McCain'");
        assertEquals(2, result.size());
        checkTerm(result.get(0), FIRST_NAME, "John and Mike", SearchTermType.VALUE);
        checkTerm(result.get(1), LAST_NAME, "McCain", SearchTermType.VALUE);
    }

    @Test
    public void testEqualsIntoQuotes() {
        List<SearchTerm<?>> result = parse("firstName='John=Mike' and lastName='McCain'");
        assertEquals(2, result.size());
        checkTerm(result.get(0), FIRST_NAME, "John=Mike", SearchTermType.VALUE);
        checkTerm(result.get(1), LAST_NAME, "McCain", SearchTermType.VALUE);
    }

    @Test
    public void testSpacesIntoQuotes() {
        List<SearchTerm<?>> result = parse("firstName='John Mike' and lastName='McCain'");
        assertEquals(2, result.size());
        checkTerm(result.get(0), FIRST_NAME, "John Mike", SearchTermType.VALUE);
        checkTerm(result.get(1), LAST_NAME, "McCain", SearchTermType.VALUE);
    }

    @Test
    public void testSpecialSymbols() {
        List<SearchTerm<?>> result = parse("'_!@firstName'='John Mike' and '_!@lastName'='McCain'");
        assertEquals(2, result.size());
        checkTerm(result.get(0), "_!@firstName", "John Mike", SearchTermType.VALUE);
        checkTerm(result.get(1), "_!@lastName", "McCain", SearchTermType.VALUE);
    }

    @Test
    public void testRange() {
        List<SearchTerm<?>> result = parse("firstName='John Mike' and age=[30,40]");
        assertEquals(2, result.size());
        checkTerm(result.get(0), FIRST_NAME, "John Mike", SearchTermType.VALUE);
        checkRangeTerm(result.get(1), "age", 30, 40);
    }


    @Test(expected = ParseException.class)
    public void testCheckThrowsExceptionOnEmptyMap() throws ParseException {
        SmartSearchQueryParser.check(Collections.emptyList(), "");
    }

    @Test(expected = IllegalStateException.class)
    public void toolkitShouldThrowIllegalStateExceptionOnInit() throws ParseException {
        SmartSearchQueryToolkit.newParser(null);
    }

    private static List<SearchTerm<?>> parse(String query) {
        try {
            return SmartSearchQueryToolkit.newParser(query).parse();
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    @Test
    public void testCaseInsensitive() {
        List<SearchTerm<?>> result = parse("firstName='John Mike' AND lastName='McCain'");
        assertEquals(2, result.size());
        checkTerm(result.get(0), FIRST_NAME, "John Mike", SearchTermType.VALUE);
        checkTerm(result.get(1), LAST_NAME, "McCain", SearchTermType.VALUE);
    }

    private void checkTerm(SearchTerm<?> term, String property, String value, SearchTermType type) {
        assertEquals(property, term.getProperty());
        assertEquals(value, term.getValue());
        assertEquals(type, term.getType());
    }

    @SuppressWarnings("unchecked")
    private void checkRangeTerm(SearchTerm<?> term, String property, long start, long end) {
        assertEquals(SearchTermType.RANGE, term.getType());
        SearchTerm<Range> rangedTerm = (SearchTerm<Range>) term;
        assertEquals(property, term.getProperty());
        assertEquals(start, rangedTerm.getValue().getStart());
        assertEquals(end, rangedTerm.getValue().getEnd());
    }

}

File: entity-browser-app/src/test/kotlin/jetbrains/xodus/browser/web/search/SmartSearchQueryParserTest.kt
package jetbrains.xodus.browser.web.search


import org.junit.Assert.assertEquals
import org.junit.Test

class SmartSearchQueryParserTest {

    @Test
    fun testSimple() {
        val result = "firstName='John' and lastName='McClane' and age!=34".parse()
        assertEquals(3, result.size.toLong())
        result[0].assertPropertyValueTerm("firstName", "John", true)
        result[1].assertPropertyValueTerm("lastName", "McClane", true)
        result[2].assertPropertyValueTerm("age", "34", false)
    }

    @Test
    fun testSingleParam() {
        val result = "firstName='John'".parse()
        assertEquals(1, result.size.toLong())
        result[0].assertPropertyValueTerm("firstName", "John", true)
    }

    @Test
    fun testWithoutQuotes() {
        val result = "firstName=John and lastName!=McClane and age=43".parse()
        assertEquals(3, result.size.toLong())
        result[0].assertPropertyValueTerm("firstName", "John", true)
        result[1].assertPropertyValueTerm("lastName", "McClane", false)
        result[2].assertPropertyValueTerm("age", "43", true)
    }

    @Test
    fun testLike() {
        val result = "firstName=John and lastName~McClane".parse()
        assertEquals(2, result.size.toLong())
        result[0].assertPropertyValueTerm("firstName", "John", true)
        result[1].assertPropertyLikeTerm("lastName", "McClane")
    }

    @Test
    fun testSingleParamsWithoutQuotes() {
        val result = "firstName=John".parse()
        assertEquals(1, result.size.toLong())
        result[0].assertPropertyValueTerm("firstName", "John", true)
    }

    @Test
    fun testEscapingQuotes() {
        val result = "firstName=John and lastName='Mc''Clane'".parse()
        assertEquals(2, result.size.toLong())
        result[0].assertPropertyValueTerm("firstName", "John", true)
        result[1].assertPropertyValueTerm("lastName", "Mc'Clane", true)
    }

    @Test
    fun testOmitAndIntoQuotes() {
        val result = "firstName!='John and Mike' and lastName='McClane'".parse()
        assertEquals(2, result.size.toLong())
        result[0].assertPropertyValueTerm("firstName", "John and Mike", false)
        result[1].assertPropertyValueTerm("lastName", "McClane", true)
    }

    @Test
    fun testEqualsIntoQuotes() {
        val result = "firstName='John=Mike' and lastName='McClane'".parse()
        assertEquals(2, result.size.toLong())
        result[0].assertPropertyValueTerm("firstName", "John=Mike", true)
        result[1].assertPropertyValueTerm("lastName", "McClane", true)
    }

    @Test
    fun testSpacesIntoQuotes() {
        val result = "firstName='John Mike' and lastName='McClane'".parse()
        assertEquals(2, result.size.toLong())
        result[0].assertPropertyValueTerm("firstName", "John Mike", true)
        result[1].assertPropertyValueTerm("lastName", "McClane", true)
    }

    @Test
    fun testSpecialSymbols() {
        val result = "'_!@firstName'='John Mike' and '_!@lastName'='McClane'".parse()
        assertEquals(2, result.size.toLong())
        result[0].assertPropertyValueTerm("_!@firstName", "John Mike", true)
        result[1].assertPropertyValueTerm("_!@lastName", "McClane", true)
    }

    @Test
    fun testRange() {
        val result = "firstName!='John Mike' and age=[30,40]".parse()
        assertEquals(2, result.size.toLong())
        result[0].assertPropertyValueTerm("firstName", "John Mike", false)
        result[1].assertPropertyRangeTerm("age", 30, 40)
    }

    @Test
    fun testLink() {
        val result = "@linkName!=MyType[3] and firstName~Jo".parse()
        assertEquals(2, result.size)
        result[0].assertLinkTerm("linkName", "MyType", 3, false)
        result[1].assertPropertyLikeTerm("firstName", "Jo")
    }

    @Test
    fun testPropNull() {
        val result = "firstName=null and lastName='null'".parse()
        assertEquals(2, result.size)
        result[0].assertPropertyValueTerm("firstName", null, true)
        result[1].assertPropertyValueTerm("lastName", "null", true)
    }

    @Test
    fun testLinkNull() {
        val result = "@user!=null and @action=MyAction[4]".parse()
        assertEquals(2, result.size)
        result[0].assertLinkTerm("user", null, null, false)
        result[1].assertLinkTerm("action", "MyAction", 4, true)
    }

    @Test(expected = jetbrains.xodus.browser.web.search.ParseException::class)
    @Throws(jetbrains.xodus.browser.web.search.ParseException::class)
    fun testCheckThrowsExceptionOnEmptyMap() {
        jetbrains.xodus.browser.web.search.SmartSearchQueryParser.check(emptyList<SearchTerm>(), "")
    }

    @Test
    fun testCaseInsensitive() {
        val result = "firstName='John Mike' AND lastName='McClane'".parse()
        assertEquals(2, result.size.toLong())
        result[0].assertPropertyValueTerm("firstName", "John Mike", true)
        result[1].assertPropertyValueTerm("lastName", "McClane", true)
    }

    private fun SearchTerm.assertPropertyLikeTerm(expectedName: String, expectedValue: String) {
        val actual = this as PropertyLikeSearchTerm
        assertEquals(expectedName, actual.name)
        assertEquals(expectedValue, actual.value)
    }

    private fun SearchTerm.assertPropertyRangeTerm(expectedName: String, expectedStart: Long, expectedEnd: Long) {
        val actual = this as PropertyRangeSearchTerm
        assertEquals(expectedName, actual.name)
        assertEquals(expectedStart, actual.start)
        assertEquals(expectedEnd, actual.end)
    }

    private fun SearchTerm.assertPropertyValueTerm(expectedName: String, expectedValue: String?, expectedEquals: Boolean) {
        val actual = this as PropertyValueSearchTerm
        assertEquals(expectedName, actual.name)
        assertEquals(expectedValue, actual.value)
        assertEquals(expectedEquals, actual.equals)
    }

    private fun SearchTerm.assertLinkTerm(expectedName: String, expectedTypeName: String?, expectedLocalId: Long?, expectedEquals: Boolean) {
        val actual = this as LinkSearchTerm
        assertEquals(expectedName, actual.name)
        assertEquals(expectedTypeName, actual.oppositeEntityTypeName)
        assertEquals(expectedLocalId, actual.oppositeEntityLocalId)
        assertEquals(expectedEquals, actual.equals)
    }
}
